### PR TITLE
feat:add the shortcut of removeif in ResourceBuilder

### PR DESCRIPTION
**Pull request by Rocksnake**
This PR is related to #4175 .
The changes included are the following：
- add shortcut of removeif in ResourceBuilder;
- optimize the code in ResourceConfiguration.

### CODE REVIEW

File: sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ResourceConfiguration.java [hunk 1]
```diff
@@ -77,9 +77,9 @@ static Resource filterAttributes(Resource resource, ConfigProperties configPrope
     Set<String> disabledKeys = new HashSet<>(configProperties.getList(DISABLED_ATTRIBUTE_KEYS));
 
     Attributes filteredAttributes =
-        resource.getAttributes().toBuilder()
+        resource.toBuilder()
             .removeIf(attributeKey -> disabledKeys.contains(attributeKey.getKey()))
-            .build();
+            .buildAttributes();
```
Comment (jack-berg) [inline sdk-extensions/autoconfigure/src/main/java/io/opentelemetry/sdk/autoconfigure/ResourceConfiguration.java:82]: No need for the `buildAttributes()` method. Instead, can replace line 84 with:
```
ResourceBuilder builder = resource.toBuilder().removeIf(attributeKey -> disabledKeys.contains(attributeKey.getKey()));
```

**Review by jack-berg — APPROVED** [review]

**Comment by Rocksnake** [comment]
@jack-berg Thanks for your approval. There is a problem with the unit test when I execute the checks. 
When I execute `./gradlew :sdk:common:test` locally, one error that `'other' has different root` was occured. I am modifying.

File: sdk/common/src/test/java/io/opentelemetry/sdk/resources/ResourceTest.java [hunk 2]
```diff
@@ -330,4 +331,45 @@ public void toBuilder() {
     assertThat(newResource.getAttribute(stringKey("foo"))).isEqualTo("val");
     assertThat(newResource.getSchemaUrl()).isEqualTo("http://example.com");
   }
+
+  @Test
+  public void removeIf() {
+    ResourceBuilder builder = Resource.getDefault().toBuilder();
+    assertThat(builder.removeIf(unused -> true)).isEqualTo(builder);
+
+    assertThat(builder.removeIf(key -> key.getKey().equals("key1"))).isEqualTo(builder);
```
Comment (jack-berg) [inline sdk/common/src/test/java/io/opentelemetry/sdk/resources/ResourceTest.java:340]: These tests need to call `build()` on `ResourceBuilder` for a proper comparison. I.e.:
```
    assertThat(builder.removeIf(key -> key.getKey().equals("key1")).build()).isEqualTo(builder.build());
```

Comment (jack-berg) [inline sdk/common/src/test/java/io/opentelemetry/sdk/resources/ResourceTest.java:340]: These tests need to call `build()` on `ResourceBuilder` for a proper comparison. I.e.:
```
    assertThat(builder.removeIf(key -> key.getKey().equals("key1")).build()).isEqualTo(builder.build());
```

File: sdk/common/src/test/java/io/opentelemetry/sdk/resources/ResourceTest.java [hunk 3]
```diff
@@ -330,4 +331,45 @@ public void toBuilder() {
...
+                        key.getKey().equals("key1") && key.getType().equals(AttributeType.STRING)))
+        .isEqualTo(builder.put("key1", true));
+
+    assertThat(
+            builder
+                .put("key1", "value1")
+                .put("key2", "value2")
+                .put("foo", "bar")
+                .removeIf(key -> key.getKey().matches("key.*")))
+        .isEqualTo(builder.put("foo", "bar"));
```
Comment (jack-berg) [inline sdk/common/src/test/java/io/opentelemetry/sdk/resources/ResourceTest.java:373]: You need to create new resources for each test case and assertion comparison for this test to be valid:
```suggestion
    ResourceBuilder builder = Resource.builder();
    assertThat(builder.removeIf(unused -> true)).isEqualTo(builder);

    assertThat(Resource.builder().removeIf(key -> key.getKey().equals("key1")).build())
        .isEqualTo(builder.build());

    assertThat(
            Resource.builder()
                .put("key1", "value1")
                .removeIf(key -> key.getKey().equals("key1"))
                .removeIf(key -> key.getKey().equals("key1"))
                .build())
        .isEqualTo(Resource.builder().build());

    assertThat(
            Resource.builder()
                .put("key1", "value1")
                .put("key1", "value2")
                .put("key2", "value2")
                .put("key3", "value3")
                .removeIf(key -> key.getKey().equals("key1"))
                .build())
        .isEqualTo(Resource.builder().put("key2", "value2").put("key3", "value3").build());

    assertThat(
            Resource.builder()
                .put("key1", "value1A")
                .put("key1", true)
                .removeIf(
                    key ->
                        key.getKey().equals("key1") && key.getType().equals(AttributeType.STRING))
                .build())
        .isEqualTo(Resource.builder().put("key1", true).build());

    assertThat(
            Resource.builder()
                .put("key1", "value1")
                .put("key2", "value2")
                .put("foo", "bar")
                .removeIf(key -> key.getKey().matches("key.*"))
                .build())
        .isEqualTo(Resource.builder().put("foo", "bar").build());
```

Comment (jack-berg) [inline sdk/common/src/test/java/io/opentelemetry/sdk/resources/ResourceTest.java:373]: You need to create new resources for each test case and assertion comparison for this test to be valid:
```suggestion
    ResourceBuilder builder = Resource.builder();
    assertThat(builder.removeIf(unused -> true)).isEqualTo(builder);

    assertThat(Resource.builder().removeIf(key -> key.getKey().equals("key1")).build())
        .isEqualTo(builder.build());

    assertThat(
            Resource.builder()
                .put("key1", "value1")
                .removeIf(key -> key.getKey().equals("key1"))
                .removeIf(key -> key.getKey().equals("key1"))
                .build())
        .isEqualTo(Resource.builder().build());

    assertThat(
            Resource.builder()
                .put("key1", "value1")
                .put("key1", "value2")
                .put("key2", "value2")
                .put("key3", "value3")
                .removeIf(key -> key.getKey().equals("key1"))
                .build())
        .isEqualTo(Resource.builder().put("key2", "value2").put("key3", "value3").build());

    assertThat(
            Resource.builder()
                .put("key1", "value1A")
                .put("key1", true)
                .removeIf(
                    key ->
                        key.getKey().equals("key1") && key.getType().equals(AttributeType.STRING))
                .build())
        .isEqualTo(Resource.builder().put("key1", true).build());

    assertThat(
            Resource.builder()
                .put("key1", "value1")
                .put("key2", "value2")
                .put("foo", "bar")
                .removeIf(key -> key.getKey().matches("key.*"))
                .build())
        .isEqualTo(Resource.builder().put("foo", "bar").build());
```

File: sdk/common/src/test/java/io/opentelemetry/sdk/resources/ResourceTest.java [hunk 4]
```diff
@@ -330,4 +331,45 @@ public void toBuilder() {
...
+    assertThat(builder.removeIf(unused -> true)).isEqualTo(builder);
+    assertThat(Resource.builder().removeIf(key -> key.getKey().equals("key1")).build())
+        .isEqualTo(builder.build());
+    assertThat(
+            Resource.builder()
+                .put("key1", "value1")
+                .removeIf(key -> key.getKey().equals("key1"))
+                .removeIf(key -> key.getKey().equals("key1"))
+                .build())
+        .isEqualTo(Resource.builder().build());
```
Comment (anuraaga) [inline sdk/common/src/test/java/io/opentelemetry/sdk/resources/ResourceTest.java:347]:
```suggestion
    assertThat(Resource.builder().removeIf(unused -> true)).isEqualTo(Resource.empty());
    assertThat(Resource.builder().removeIf(key -> key.getKey().equals("key1")).build())
        .isEqualTo(Resource.empty());
    assertThat(
            Resource.builder()
                .put("key1", "value1")
                .removeIf(key -> key.getKey().equals("key1"))
                .removeIf(key -> key.getKey().equals("key1"))
                .build())
        .isEqualTo(Resource.empty());
```

**Review by anuraaga — COMMENTED** [review]
Thanks, just last small cleanup

File: sdk/common/src/test/java/io/opentelemetry/sdk/resources/ResourceTest.java (hunk 3 above)
Comment (Rocksnake) [reply sdk/common/src/test/java/io/opentelemetry/sdk/resources/ResourceTest.java]: Great thanks! I have corrected it as requested.

File: sdk/common/src/test/java/io/opentelemetry/sdk/resources/ResourceTest.java (hunk 4 above)
Comment (Rocksnake) [reply sdk/common/src/test/java/io/opentelemetry/sdk/resources/ResourceTest.java]: Great thanks! I have corrected it as requested.

**Review by anuraaga — APPROVED** [review]

**Comment by anuraaga** [comment]
Thanks!

**Pull request merged**